src/mainwin.cc

```diff
@@ -2013,6 +2017,164 @@ void MainWindow::csgRender()
+	static unsigned int printCounter=0;
+	
+	setCurrentOutput();
+	PRINT("3D Printing...");
+	
+	unsigned int dim = 3;
+	
+	QUrl partUrl;
+	
+	//Where we hold our temporary stl file name:
+	char tempStlFileName[L_tmpnam];
+	
+	setCurrentOutput();
+
+	//Make sure we can export:
+	if (! canExport(dim))
+	{
+		PRINT("Cannot 3D Print due to errors.");
+		return;
+	}
+	
+	//Ccreate a temporary file name valid on all systems:
+	QString exportFilename=QString(std::tmpnam(tempStlFileName));
+	
+	//Render the stl to a temporary file:
+	exportFileByName(this->root_geom, FileFormat::STL,
+		exportFilename.toLocal8Bit().constData(),
+		exportFilename.toUtf8());
+	
+	//Create a name that the order process will use to refer to the file. Base it off of the
+	//project name
+	QString userFacingName="unsaved.stl";
+	if (!this->fileName.isEmpty()) {
+		QString fileBasename = QFileInfo(this->fileName).baseName();
+		userFacingName = fileBasename.append("_").append(QString::number(printCounter++)).append(".stl");
+	}
+	
+	//Upload the file to the 3D Printing server and get the corresponding url to see it.
+	//The result is put in partUrl.
+	bool uploadWorked=uploadStlAndGetPartUrl(exportFilename, userFacingName, partUrl);
+	setCurrentOutput();
+
+	//Check the result:
+	if (! uploadWorked)
+	{
+		PRINT("An error occured while contacting the 3D Print API.");
+		return;
+	}
+	
+	//PRINT("partUrl:");
+	//PRINT(partUrl);
+	
+	//Open the url in the default browser:
+	QDesktopServices::openUrl ( partUrl );
+}
+
+//This function uploads an stl to the 3D printing API endpoint and returns a url that, 
+//when accessed, will show the stl file as a part that can be configured and added to the 
+//shopping cart.  Returns True if successful.
+//
+//Inputs:
+//    exportFilename  - The path to the temporary file that has the stl export in it.
+//    userFacingName  - Then name we should give the file when it is uploaded for the order process.
+//Outputs:
+//    partUrl         - The resulting url to go to next to continue the order process.
+bool MainWindow::uploadStlAndGetPartUrl(const QString & exportFilename, const QString &userFacingName, QUrl &partUrl)
+{
+	setCurrentOutput();
+	
+	//Create a request:
+	QNetworkRequest request(QUrl("https://print.openscad.org/api/v1/part-upload/"));
+	
+	//Set the content header:
+	request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
+	
+	//Get the file name from the file path:
+	QString fileNameBase=QFileInfo(exportFilename).fileName();
+	
+	//Create the request:
+	QJsonObject jsonInput;
+	
+	//Start building the json request:
+	jsonInput.insert("fileName", userFacingName);
+	
+	//Read our stl file:
+	QFile file(exportFilename);
+	if (!file.open(QIODevice::ReadOnly))
+	{
+		PRINT("Unable to open exported stl file.");
+		return 0;
+	}
+	
+	//Convert it to base 64:
+	QString stlFileB64=file.readAll().toBase64();
+	
+	//Base 64-encoded file contents:
+	jsonInput.insert("file", stlFileB64.toLocal8Bit().constData());
```

Comment:
This seems to confuse older Qt (on Travis / Trusty). Base64 encoded string should be safe without that `.toLocal8Bit().constData()`.

Reply:
Good catch!  :)  Removed.

Reply:
I presume the API will document which [Base64](https://en.wikipedia.org/wiki/Base64#Implementations_and_history)? QT happens to use [RFC 2045](http://doc.qt.io/archives/qt-4.8/qbytearray.html#toBase64).